Shipping Labels M2: display the Package List

pmusolino opened the pull request:
Closes #3958  

## Description
In this PR I implemented the screen for displaying the Package List in the Shipping Labels.
All the components are implemented in SwiftUI. The flow was implemented using a view that contains another subview. Wanting to be more specific, `ShippingLabelAddNewPackage` is the container of the reusable `ShippingLabelPackageList` view. I implemented it in this way because in the future we will use the `ShippingLabelPackageList` component for the Package Creation Flow. So, the `ShippingLabelPackageList` will be embedded in another view where the user can create a Custom Package or select an existing package.

## Changes
- Implemented the `ShippingLabelPackageList` view that displays a list of Custom Packages and Predefined Packages. I also implemented the view model `ShippingLabelPackageListViewModel` and the tests. The view model takes care of managing the state of the view and the selected package.
- I made some small UI changes to the existing `ShippingLabelAddNewPackage` view, removing the unused (for the moment) `SegmentedView`. I also made some changes to the view model, which now takes care of the States and of fetching the package details. I added the unit tests also for this view model.
- I implemented a new reusable view in SwiftUI, called `SelectableItemRow` that displays a title, a subtitle, and the checkmark on the left side when selected.

## Testing
1. Launch the app in debug mode (because of the feature flag).
2. Open an order detail.
3. Press the button "Create Shipping Label".
4. Press the button "Continue" in Create Shipping Label form, under the **Ship From** cell. Confirm the address.
5. Press the button "Continue" in Create Shipping Label form, under the **Ship To** cell. Confirm the address.
6. Press the button "Continue" in Create Shipping Label form, under the **Package Details** cell.
7. You should see the new screen that shows the Package Details screen and all the items that will be shipped.
8. Press the "Package selected" row. -> You should see the new screen with the list of the existing packages, with Custom Packages on top if any.

For the moment the `error` and the `syncing` states are not managed and will be part of other PRs. https://github.com/woocommerce/woocommerce-ios/issues/3965 https://github.com/woocommerce/woocommerce-ios/issues/3966 Also the handle of the value will be part of another PR https://github.com/woocommerce/woocommerce-ios/issues/3964

| Package List (Light mode)             |  Package List (Dark mode)  |
:-------------------------:|:-------------------------:
![Simulator Screen Shot - iPhone 12 Pro - 2021-04-15 at 14 26 30](https://user-images.githubusercontent.com/495617/114868993-cbb3b580-9df6-11eb-8b1d-315731731573.png) | ![Simulator Screen Shot - iPhone 12 Pro - 2021-04-15 at 14 27 19](https://user-images.githubusercontent.com/495617/114869003-ce160f80-9df6-11eb-8397-8a2195f3e391.png)



Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

pmusolino committed:
feat: now the ListHeaderView support the multiline text

pmusolino committed:
feat: implemented basic struct for ShippingLabelServicePackagesView in SwiftUI

pmusolino committed:
feat: added typealias ShippingLabelPredefinedOption in Yosemite Model

pmusolino committed:
feat: added ShippingLabelServicePackagesViewModel and ShippingLabelServicePackages

pmusolino committed:
feat: added `siteID` property  in ShippingLabelPackageDetails view model

pmusolino committed:
feat: sync package details

pmusolino committed:
feat: implemented SelectableItemRow SwiftUI component

pmusolino committed:
feat: ShippingLabelServicePackages

pmusolino committed:
feat: ShippingLabelServicePackages improvements

pmusolino committed:
fix: typo

pmusolino committed:
feat: exposed Networking.ShippingLabelStoreOptions to Yosemite

pmusolino committed:
feat: implemented ShippingLabelServicePackages and view model

pmusolino committed:
refactor: ShippingLabelAddNewPackage and ViewModel

pmusolino committed:
feat: package selection

pmusolino committed:
feat: ShippingLabelAddNewPackageViewModel unit tests

pmusolino committed:
refactor: renamed ShippingLabelServicePackages to ShippingLabelPackageList and ShippingLabelServicePackagesViewModel to ShippingLabelPackageListViewModel

pmusolino committed:
feat: ShippingLabelPackageListViewModel unit tests

pmusolino committed:
feat: added dimension unit at the end of every selectable item in the Package List

pmusolino committed:
fix: removed superflous ScrollView

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Add New Package/Service Package List/ShippingLabelPackageList.swift:
```diff
@@ -0,0 +1,92 @@
+            /// Custom Packages
+            ///
+            if viewModel.customPackages.count > 0 {
+                ListHeaderView(text: Localization.customPackageHeader, alignment: .left)
```
Would it make sense to use `Localization.customPackageHeader.uppercased()` here (similar to the headers for Predefined Packages below) instead of relying on the string `Localization.customPackageHeader` to be in uppercase?

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Add New Package/Service Package List/ShippingLabelPackageList.swift:
```diff
@@ -0,0 +1,92 @@
+        let storeOptions = ShippingLabelStoreOptions(currencySymbol: "$", dimensionUnit: "in", weightUnit: "oz", originCountry: "US")
+
+        let customPackages = [
+            ShippingLabelCustomPackage(isUserDefined: true, title: "Box", isLetter: true, dimensions: "3 x 10 x 4", boxWeight: 10, maxWeight: 11),
```
nit: For consistency/readability, these parameters could each be on their own line (like the other `ShippingLabelCustomPackage`s below).

rachelmcr commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/SelectableItemRow.swift:
```diff
@@ -0,0 +1,49 @@
+            Spacer()
+        }
+        .padding([.top, .bottom], Constants.hStackPadding)
+        .frame(minHeight: Constants.height)
```
I noticed that when `SelectableItemRow` is used, only the part of the row with text in it is tappable.

Based on [this article](https://www.hackingwithswift.com/quick-start/swiftui/how-to-control-the-tappable-area-of-a-view-using-contentshape), it looks like we can make the whole row tappable by adding `contentShape()`. Adding something like this to the `HStack` should work:

```swift
.frame(minHeight: Constants.height)
.contentShape(Rectangle())
```

pmusolino replied:
Good catch! Fixed 👍

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/Add New Package/ShippingLabelAddNewPackage.swift:
```diff
@@ -1,32 +1,31 @@
 struct ShippingLabelAddNewPackage_Previews: PreviewProvider {
     static var previews: some View {
-        ShippingLabelAddNewPackage()
+        ShippingLabelAddNewPackage(siteID: 1234)
```
I can't think of a straightforward way to do this, but do you think it'd be possible to fake the site in a way that will make this SwiftUI preview work? As it is, the preview just shows a blank screen. It isn't critical, but it'd be nice to have the preview working.

pmusolino replied:
I thought about this thing, but this would involve providing a method to access data from the outside (therefore without going through the Yosemite layer), and which would be used only for the preview, and I didn't particularly like the solution. 🤔

rachelmcr reviewed (commented):
This is really nice! One thing I'm not sure about, though, is the container view:

> Wanting to be more specific, ShippingLabelAddNewPackage is the container of the reusable ShippingLabelPackageList view. I implemented it in this way because in the future we will use the ShippingLabelPackageList component for the Package Creation Flow.

In the designs, the package list is displayed on a screen titled “Package Selected.” In this implementation, the screen is titled “Add New Package” (which is part of the creation flow design). Should we have two different container views that we can use depending on which flow we're in?

I also noticed that the screen title "Add New Package" is also showing up in the very bottom left (see my screenshots); do you know why that's happening?

I have a few other comments below, but overall this is working really well for me. A few screenshots from testing with different package lists:

Custom & Predefined packages | Predefined only | Custom only
-|-|-
![Simulator Screen Shot - iPhone 11 Pro - 2021-04-16 at 16 00 00](https://user-images.githubusercontent.com/8658164/115045982-e78a8a80-9ece-11eb-9a41-7ecf51811bd2.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-04-16 at 16 08 46](https://user-images.githubusercontent.com/8658164/115045990-e8bbb780-9ece-11eb-8af1-ec8d2a100a41.png)|![Simulator Screen Shot - iPhone 11 Pro - 2021-04-16 at 16 09 51](https://user-images.githubusercontent.com/8658164/115046002-e9ece480-9ece-11eb-97dd-2961c2c428fd.png)

pmusolino committed:
feat: code style improvements + fixed tap on SelectableItemRow

pmusolino committed:
refactor: renamed ShippingLabelAddNewPackage to ShippingLabelPackageSelected

pmusolino commented:
Thank you @rachelmcr for the review! I fixed all the things that you discovered 👍 

> In the designs, the package list is displayed on a screen titled “Package Selected.” In this implementation, the screen is titled “Add New Package” (which is part of the creation flow design). Should we have two different container views that we can use depending on which flow we're in?

You are right, I fixed it and renamed the `ShippingLabelAddNewPackage` view as `ShippingLabelPackageSelected`. 

> I also noticed that the screen title "Add New Package" is also showing up in the very bottom left (see my screenshots); do you know why that's happening?

I tried to investigate why this happens, but I wasn't able to find a quick solution, so I opened an issue here https://github.com/woocommerce/woocommerce-ios/issues/3995 I'll fix it before shipping the M2.

Can you do a second check, please? 🙇

rachelmcr reviewed (approved):
Looks great! 🎉   :shipit:
